File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/Intonation/IntonationViewModel.swift
import Foundation
import Combine
import VoxaTrace

/// ViewModel for intonation analysis using Accura.
///
/// ## VoxaTrace Integration
/// ```swift
/// // 1. Decode audio
/// let audioData = SonixDecoder.decode(path: filePath)
///
/// // 2. Extract pitch contour
/// let extractor = PitchDetection.createContourExtractor()
/// let contour = extractor.extract(audio: samples, sampleRate: sampleRate)
/// extractor.release()
///
/// // 3. Analyze intonation against EQ and JI
/// let eqResult = Accura.analyzePitching(contour: contour, tonicHz: 146.83, intonationSystem: .eq)
/// let jiResult = Accura.analyzePitching(contour: contour, tonicHz: 146.83, intonationSystem: .ji)
///
/// // 4. Calculate scores (only when analysis succeeded — ADR-022)
/// let eqScore = eqResult.error == nil ? Accura.calculateScore(result: eqResult) : nil
/// let jiScore = jiResult.error == nil ? Accura.calculateScore(result: jiResult) : nil
/// ```
@MainActor
final class IntonationViewModel: ObservableObject {

    // MARK: - Published State

    @Published private(set) var eqResult: IntonationAnalysisResult?
    @Published private(set) var jiResult: IntonationAnalysisResult?
    @Published private(set) var eqScore: PitchingScore?
    @Published private(set) var jiScore: PitchingScore?
    @Published private(set) var isAnalyzing = false

    // MARK: - Actions

    func analyzeOffline() {
        isAnalyzing = true
        eqResult = nil
        jiResult = nil
        eqScore = nil
        jiScore = nil

        Task {
            guard let audioURL = Bundle.main.url(forResource: "Alankaar 01_voice", withExtension: "m4a"),
                  let audioData = SonixDecoder.decode(path: audioURL.path) else {
                await MainActor.run { isAnalyzing = false }
                return
            }

            // Extract pitch contour
            let extractor = PitchDetection.createContourExtractor()
            let contour = extractor.extract(audio: audioData.samples, sampleRate: audioData.sampleRate)
            extractor.release()

            let tonicHz: Float = 146.83

            // Analyze against Equal Temperament
            let eq = Accura.analyzePitching(
                contour: contour,
                tonicHz: tonicHz,
                intonationSystem: .eq,
                noteLabelTradition: .carnatic
            )

            // Analyze against Just Intonation
            let ji = Accura.analyzePitching(
                contour: contour,
                tonicHz: tonicHz,
                intonationSystem: .ji,
                noteLabelTradition: .carnatic
            )

            // Calculate scores (only when analysis succeeded — ADR-022)
            let eqSc = eq.error == nil ? Accura.calculateScore(result: eq) : nil
            let jiSc = ji.error == nil ? Accura.calculateScore(result: ji) : nil

            await MainActor.run {
                eqResult = eq
                jiResult = ji
                eqScore = eqSc
                jiScore = jiSc
                isAnalyzing = false
            }
        }
    }
}

File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/intonation/viewmodel/IntonationViewModel.kt
package com.musicmuni.voxatrace.demo.sections.intonation.viewmodel

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.musicmuni.voxatrace.accura.Accura
import com.musicmuni.voxatrace.accura.model.IntonationAnalysisResult
import com.musicmuni.voxatrace.accura.model.IntonationSystem
import com.musicmuni.voxatrace.accura.model.NoteLabelTradition
import com.musicmuni.voxatrace.accura.model.PitchingScore
import com.musicmuni.voxatrace.sonix.SonixDecoder
import com.musicmuni.voxatrace.tona.PitchDetection
import io.github.aakira.napier.Napier
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream

/**
 * ViewModel for Intonation Analysis - pitch accuracy against EQ and JI tuning systems.
 *
 * APIs demonstrated:
 * - Accura.analyzePitching() for intonation analysis
 * - Accura.calculateScore() for scoring
 * - PitchDetection.createContourExtractor() for pitch extraction
 * - SonixDecoder.decode() for audio decoding
 */
class IntonationViewModel : ViewModel() {

    // Published state
    private val _eqResult = MutableStateFlow<IntonationAnalysisResult?>(null)
    val eqResult: StateFlow<IntonationAnalysisResult?> = _eqResult.asStateFlow()

    private val _jiResult = MutableStateFlow<IntonationAnalysisResult?>(null)
    val jiResult: StateFlow<IntonationAnalysisResult?> = _jiResult.asStateFlow()

    private val _eqScore = MutableStateFlow<PitchingScore?>(null)
    val eqScore: StateFlow<PitchingScore?> = _eqScore.asStateFlow()

    private val _jiScore = MutableStateFlow<PitchingScore?>(null)
    val jiScore: StateFlow<PitchingScore?> = _jiScore.asStateFlow()

    private val _isAnalyzing = MutableStateFlow(false)
    val isAnalyzing: StateFlow<Boolean> = _isAnalyzing.asStateFlow()

    // Actions

    fun analyzeOffline(context: Context) {
        viewModelScope.launch {
            _isAnalyzing.value = true
            _eqResult.value = null
            _jiResult.value = null
            _eqScore.value = null
            _jiScore.value = null

            try {
                val audioFile = withContext(Dispatchers.IO) {
                    copyAssetToFile(context, "Alankaar 01_voice.m4a")
                }

                val audioData = withContext(Dispatchers.IO) {
                    SonixDecoder.decode(audioFile.absolutePath)
                }

                if (audioData == null) {
                    Napier.e("Failed to decode audio file")
                    _isAnalyzing.value = false
                    return@launch
                }

                // Extract pitch contour via tona
                val extractor = PitchDetection.createContourExtractor()
                val contour = withContext(Dispatchers.IO) {
                    extractor.extract(audioData.samples, audioData.sampleRate)
                }
                extractor.release()

                // Analyze against Equal Temperament
                val eqAnalysis = withContext(Dispatchers.IO) {
                    Accura.analyzePitching(
                        contour,
                        tonicHz = 146.83f,
                        intonationSystem = IntonationSystem.EQ,
                        noteLabelTradition = NoteLabelTradition.CARNATIC
                    )
                }
                _eqResult.value = eqAnalysis

                if (eqAnalysis != null) {
                    _eqScore.value = Accura.calculateScore(eqAnalysis)
                }

                // Analyze against Just Intonation
                val jiAnalysis = withContext(Dispatchers.IO) {
                    Accura.analyzePitching(
                        contour,
                        tonicHz = 146.83f,
                        intonationSystem = IntonationSystem.JI,
                        noteLabelTradition = NoteLabelTradition.CARNATIC
                    )
                }
                _jiResult.value = jiAnalysis

                if (jiAnalysis != null) {
                    _jiScore.value = Accura.calculateScore(jiAnalysis)
                }

            } catch (e: Exception) {
                Napier.e("Intonation analysis failed", e)
            } finally {
                _isAnalyzing.value = false
            }
        }
    }

    private fun copyAssetToFile(context: Context, assetName: String): File {
        val file = File(context.cacheDir, assetName)
        if (!file.exists()) {
            context.assets.open(assetName).use { input ->
                FileOutputStream(file).use { output ->
                    input.copyTo(output)
                }
            }
        }
        return file
    }
}
